### ISSUE TITLE
Security Vulnerability exposed via Trivy Scan due to dependancy on `System.Drawing.Common`

### ISSUE THREAD

**Issue by andy-g** [fluentmigrator/fluentmigrator#1684]
**Is your feature request related to a problem? Please describe.**
I recently ran a Trivy image scan against an image using FluentMigrator, and it flagged a Critical Vulnerability related to `System.Drawing.Common` Library (specifically version 4.7.0).

<img width="1132" alt="image" src="https://github.com/fluentmigrator/fluentmigrator/assets/557065/6980018b-83f5-4d12-83c9-6bd3629c6022">

I couldn't clearly see where this package is included in FluentMigrator, but if I removed FluentMigrator and rebuilt my application, the vulnerability is no longer found and `System.Drawing.Common` is no longer included.

https://avd.aquasec.com/nvd/2021/cve-2021-24112/

**Describe the solution you'd like**
This package should be updated to a patched version, but ideally shouldn't even be included. Im building an API so the idea of having `System.Drawing.Common` included feels quite unnecessary. In addition to that, I'm actually building a linux based container image, and from what I can see, it's not even supported on non windows environments. There does seem to be a number of other `System.*` namespace packages included that I'm not sure are actually required either.

**Additional context**
The only reference to `System.Drawing` I found is here, but maybe with reflection etc., that’s enough to pull that package in:
[https://github.com/fluentmigrator/fluentmigrator/blob/028340e1e670321d9e19619dccaf[…]c/FluentMigrator.Abstractions/Infrastructure/ErrorMessages.resx](https://github.com/fluentmigrator/fluentmigrator/blob/028340e1e670321d9e19619dccaf7735deeaae2e/src/FluentMigrator.Abstractions/Infrastructure/ErrorMessages.resx#L21)

**Is this a feature you'd like to submit a PR for?**
I couldn't find where this was included, so not too sure where it should be removed, or what it's been included for.

**Comment by andy-g** [comment]
I also would have expected to be able to just include a reference to `FluentMigrator.Runner.Postgres` (I'm doing migrations on a Postgres DB):
```xml
<PackageReference Include="FluentMigrator.Runner.Postgres" Version="3.3.2" />
```
...but I've needed need add references to multiple packages is that expected (it seems like this is what your documentation suggests):
```xml
<PackageReference Include="FluentMigrator" Version="3.3.2" />
<PackageReference Include="FluentMigrator.Extensions.Postgres" Version="3.3.2" />
<PackageReference Include="FluentMigrator.Runner" Version="3.3.2" />
``` 

I'm also seeing now that even though I assume I should only need the postgres extension, the other extensions & runners are also included in the build output (`Oracle`, `SqlServer`, `DB2`, `Firebird`, `Hana` etc.).

Unless I'm missing something, it'd be ideal to simply include a single package (in my case I assume that'd be `FluentMigrator.Runner.Postgres`), that should then include the core `FluentMigrator` & `FluentMigrator.Runner` packages, but should exclude all the other DB runners & extensions, as well as any other unnecessary packages such as `System.Drawing.Common`.

**Comment by razaqK** [comment]
I think there is need to allow the use of needed runners instead of having dependencies of runner that is not needed in a project. 

Looking at the dependencies, the runner responsible for the `System.Drawing.Common` is `FluentMigrator.Runner.SqlServerCe`. 

See the dependencies tree below
```
FluentMigrator.Runner.SqlServerCe
  -> System.Security.Permissions 
       -> `System.Windows.Extensions 
         -> System.Drawing.Common 
```   
     
Also the package `System.Drawing.Common` only support the window platform, see here https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/system-drawing-common-windows-only


A workaround to the issue is to explicitly install version 5.0.3 of `System.Drawing.Common` package.
```
<PackageReference Include="System.Drawing.Common" Version="5.0.3" />
```

Same issue was reported on [StackExchange.Redis](https://github.com/StackExchange/StackExchange.Redis/issues/2283) And [Microsoft.Data.SqlClient](https://github.com/dotnet/SqlClient/issues/2040)

**Comment by jzabroski** [comment]
In the main branch, CE is removed. 4.0 will not have it. Just need to wrap up 4.0. Unfortunately in the middle of a major project professionally going live in two weeks, so I'm not focused on shipping 4.0 just yet.

**Comment by robertcoltheart** [comment]
@andy-g While you wait for 4.0, you can use this https://www.nuget.org/packages/FluentMigratorBridge which lets you use just a single database runner instead of referencing them all. It might resolve your issue in the short-term.

**Comment by andy-g** [comment]
Thanks @robertcoltheart that looks useful. We managed to work around the vulnerability using @razaqK's suggestion of referencing a patched version of `System.Drawing.Common`, but we're obviously still pulling in a whole set of unneeded packages, which your package would resolve.

@jzabroski please do let me know if there's anything I can help with to get 4.0 over the line.

**Comment by jzabroski** [comment]
I think I am very close to tagging the 4.0.0 branch now off the master branch.  Just folded in a bunch of stuff from the develop branch including Snowflake support.  All vulnerabilities are addressed and all dependencies are now compile-at-nuget-head.

**Comment by jzabroski** [comment]
Closing as SqlServerCe support was removed in #1229

**Comment by andy-g** [comment]
Thanks @jzabroski, I assume your recommendation would be to just wait till 4.0 is available on nuget right? And will simply installing the 4.0 version of the `FluentMigrator.Runner.Postgres` module include any additional dependancies it has - such as the base `FluentMigrator` package?

**Comment by jzabroski** [comment]
Hi @andy-g

Main remaining item blocking release right now is compiling release notes. Could you please help me with that?

It's unreal but I'm actually sick again and want to push this out.

In terms of development improvements, runners are now modular but dotnet-fm (Fluentmigrator.Dotnet.Cli) is not. To be transparent and somewhat controversial, dotnet-fm probably should be deprecated given a lot of people deploy it on production machines and that means installing the dotnet sdk on the production machine. The workaround is to "lift" all setup logic in Fluentmigrator.dotnet.cli into a FluentMigratorTaskRunnerBase abstract base class and allow system actors (e.g. You as a consumer, or Fluentmigrator.Runner.Postgres as a single runner) to override DI defaults as necessary. Such a template would avoid chaos of people doing different mistakes configuring DI.

**Comment by andy-g** [comment]
Hi @jzabroski , happy to try help with release notes, but also not really sure where to get started, I can take a look through previous release notes and see what you did there, but was also struggling to find a v4.0 PR or branch, which I assumed would be the right place to try determine what changes have been made in this release.

**Comment by andy-g** [comment]
I was also wondering if you're feeling that `Fluentmigrator.Dotnet.Cli` isn't really v 4.0 ready, and may actually be better to deprecate, is it not an option to simply not include that in the v4.0 release, it's then not officially deprecated yet, but anyone that is needing it will need to stick with v3 for now.

I think if you're wanting to use a bundled cli rather than write you're own console app to do the migration (or include the migrations as part of your application startup), I think it's a reasonable to bundle all the dependencies in that CLI for any DB that the CLI may need to access, but when explicitly including a package, it's ideal to pull in only the required dependancies.

**Comment by andy-g** [comment]
Yes, that sort of makes sense, but I've been out of the windows world for decades :), came back to dotnet as it transitioned to dotnet core, and now developing on Mac and running entirely on linux based containers - so maybe not representative of the majority of the Fluentmigrator users.

**Comment by jzabroski** [comment]
Hope about this plan of attack: We release what we have, and you help with a PR for the improved task runner workflow.

Releasing what we have mitigates security vulnerabilities, and addresses functionality users desire (smaller exe size, newer library dependencies, some improved features for postgres and mysql).  In terms of release notes, I had compiled a master list of everything that was different from the develop branch to the main branch awhile ago as part of putting a merge together.  https://docs.google.com/spreadsheets/d/1dd8bD5pTznS85auIzTCCXF44Ol_gcjpdXJhr6cB6fDc/edit?usp=sharing

I think we want things where Build Related = False, Needs Merging = Yes.  That looks like 44 items to me.  

After that, it is everything since the last tagged release.  This is a decent amount of work.  I would guess 4-8 hours to compile accurate release notes.

**Comment by andy-g** [comment]
Hi @jzabroski I'm so sorry I've dropped off the radar, works just been rather crazy lately, I haven't forgotten this so will reach out if & when I have some capacity.

**Comment by jzabroski** [comment]
You're fine. I'm going to complete the release notes this weekend.

Enjoy your holidays.

**Comment by jzabroski** [comment]
@andy-g I totally forgot GitHub has a new, automated Draft Release Notes feature.  This was wayyyy easier than the past few years. Wow. As a maintainer, that button click is a life saver. 5.0.0 is released.  I skipped 4.0.0 because technically the develop branch has been posting 4.0 versions for awhile, so for some people it might be awkward to call it 4.0.0.  To prevent package managers from being confused, and because this now supports .NET 7 and 8, I named it 5.0.0
